Add Jenny's Helm chart

claudusd opened the pull request:
This PR add the Helm chart from Jenny. A Django base app.

claudusd committed:
🙈(ignore) add gitignore file

Ingore the tgz build with the command helm package.

claudusd committed:
🔧(jenny) add anotations on the ingress

This annotations need to be add to use the
cert-manager for certificate request.

claudusd committed:
🔒️(jenny) required host for the ingress

The host is not mandatory in the ingress definition
but for an HTTP app it's a best pratice to use an
host instead of an IP. And the host is mandatory for
HTTPS.

claudusd committed:
✨(jenny) add tls support for the ingress

Add the possibility to enable the HTTPS
and to set the certificate name.

claudusd committed:
🔧(jenny) remove securityContext

The Jenny image is build to ensure the user
cannot write in the app dir. Allow to change the
running user is not required.

claudusd committed:
🔧(jenny) add nginx in the pod

Django required to collect the static and
service with a HTTP server. We add an Nginx
container in the pods and add an init container
to collect this statics. And use an ephemeral
volume for static instead of a persistent one.

claudusd committed:
🐛(jenny) fix job creation

Fix when create a list of job in the same template and
ensure to run the job after the install to avoid missing
resource.

claudusd committed:
✨(jenny) allow to add env from secret

This allow from the values to add environment
variable from secret to configure Jenny's app.

claudusd committed:
🙈(jenny) add helmignore

Ensure to ingore previous helm package
in the next helm package.

claudusd committed:
🔧(jenny) use proper check for kube probes

For the Django app we use the hearbeat add by the
lib dockerflow for the liveness and readiness probes
Use Nginx status page for liveness probe.
And check migration for the startup probe.

wilbrdt commented on jenny/templates/deployment.yaml:
```diff
@@ -0,0 +1,123 @@
+            {{- include "jenny.envs" . | nindent 12 }}
+            - name: "UWSGI_PORT"
+              value: "{{ .Values.django.port }}"
+          {{- if .Values.persistence.enabled }}
```
I would align this with its closing `{{- end }}`, for readability. WDYT?

claudusd replied:
I remove this part.

wilbrdt commented on jenny/templates/deployment.yaml:
```diff
@@ -0,0 +1,123 @@
+              port: {{ .Values.django.port }}
+              httpHeaders:
+                - name: "Host"
+                  value: "{{ .Values.ingress.host}}"
```
I think ingress should have a default `host` value

wilbrdt commented on jenny/templates/_helpers.tpl:
```diff
@@ -0,0 +1,97 @@
+service: app
+{{- end }}
+
+{{- define "jenny.envs" -}}
```
nit:
```suggestion
{{/*
Environment variables
*/}}
{{- define "jenny.envs" -}}
```

wilbrdt commented on jenny/templates/_helpers.tpl:
```diff
@@ -0,0 +1,97 @@
+{{- end }}
+{{- end }}
+
+{{- define "django.imagePullSecrets" -}}
```
nit:
```suggestion
{{/*
ImagePullSecrets
*/}}
{{- define "django.imagePullSecrets" -}}
```

wilbrdt commented on jenny/templates/jobs.yaml:
:1st_place_medal: nice work, thanks! :pray:

wilbrdt commented on jenny/templates/jobs.yaml:
```diff
@@ -0,0 +1,44 @@
+    {{- if $.Values.commonLabels }}
+    {{ toYaml $.Values.commonLabels | nindent 4 }}
+    {{- end }}
+    app.kubernetes.io/component: jenny
```
What is this label used for?
If useful, could we use a helper definition to specify it?

wilbrdt commented on jenny/values.yaml:
```diff
@@ -0,0 +1,85 @@
+                  - "jenny-app"
+          topologyKey: kubernetes.io/hostname
+
+autoscaling:
```
Should we provide a `HorizontalPodAutoscaler`?

wilbrdt commented on jenny/templates/_helpers.tpl:
```diff
@@ -0,0 +1,97 @@
+  value: "{{ .Values.django.db.host }}"
+- name: "DB_PORT"
+  value: "{{ .Values.django.db.port }}"
+{{- range $key, $val := .Values.env.secret }}
```
When linting this chart, a warning seems to be raised because of this line. It is resolved when adding an empty default value as so:
```
env:
  secret: []
```

wilbrdt commented on jenny/values.yaml:
```diff
@@ -0,0 +1,85 @@
+podAnnotations: {}
+
+service:
+  port: 8080
```
This service port isn't used by any template, should it be removed?

wilbrdt commented on jenny/templates/configmap.yaml:
```diff
@@ -0,0 +1,103 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: "{{ template "jenny.fullname" . }}-nginx"
```
We're missing common labels and annotations here. Is that OK?

wilbrdt reviewed (commented):
Nice work, really good model!
A few general remarks:
- some files are missing an EOF newline
- labels and annotations seem a bit inconsistent, which makes them harder to use

wilbrdt commented on jenny/templates/deployment.yaml:
```diff
@@ -0,0 +1,123 @@
+            items:
+              - key: config
+                path: nginx.conf
+            name: "{{ .Chart.Name }}-nginx"
```
```suggestion
            name: "{{ template "jenny.fullname" . }}-nginx"
```

wilbrdt commented on jenny/templates/deployment.yaml:
```diff
@@ -0,0 +1,123 @@
+            - "python"
+            - "manage.py"
+            - "collectstatic"
+            - "--no-input"
```
It seems that the Django SECRET_KEY is needed to execute this command.
We should probably add the env variables to the initContainer. WDYT?

wilbrdt commented on jenny/templates/ingress.yaml:
```diff
@@ -0,0 +1,38 @@
+    {{- end }}
+spec:
+  ingressClassName: {{ .Values.ingress.class_name }}
+  {{- if $.Values.ingress.tls.enable }}
```
```suggestion
  {{- if $.Values.ingress.tls.enabled }}
```

wilbrdt commented on jenny/templates/ingress.yaml:
```diff
@@ -0,0 +1,38 @@
+  tls:
+  - hosts:
+      - {{ $.Values.ingress.host }}
+    secretName: {{ $.Values.ingress.tls.certificateSecretName }}
```
```suggestion
      secretName: {{ $.Values.ingress.tls.certificateSecretName }}
```

claudusd replied:
When I check the [spec](https://kubernetes.io/docs/concepts/services-networking/ingress/#tls) the indent is correct.

claudusd committed:
fixup! 🔧(jenny) add nginx in the pod

claudusd committed:
fixup! ✨(jenny) allow to add env from secret

claudusd committed:
fixup! ✨(jenny) add tls support for the ingress

wilbrdt commented on jenny/templates/configmap.yaml:
```diff
@@ -0,0 +1,56 @@
+          location /__status__ {
+            stub_status;
+            allow 127.0.0.1;
+            deny all;
```
```suggestion
```

claudusd committed:
fixup! 🔧(jenny) use proper check for kube probes

claudusd committed:
🔧(jenny) add resources for django and nginx

It's a best pratice to defined request resources and
limit resources. To ensure to have enought resource when
pod start and to limit memory and cpu usage.

claudusd committed:
♻️(jenny) refaco labels and add validation

Improve the metadata, selectors and add validation.

wilbrdt reviewed (approved):
:ship:
